### ISSUE TITLE
Consider merging scripts repository

### ISSUE THREAD

**Issue by kofalt** [scitran/scitran#2]

**Comment by gsfr** [comment]
Do you mean merging into this repo?

**Comment by kofalt** [comment]
My proposal would be that any repository that cannot stand on its own would be merged into this repository. There are ways we can do this that would preserve all history, as well.

In particular, I would merge at least /api; /data or /reaper perhaps. Up for discussion.

**Comment by Eric89GXL** [comment]
+1 from my end. Drawbacks I've encountered based on the current approach:

* In my experience, it is a non-standard setup (e.g., `sklearn`, `scipy`, `nibabel`, etc. are all integrated), so it adds learning overhead, and maintenance overhead in the sense that improvements to other repos methods cannot be as easily translated.
* I currently have 3 PRs open to get the MEEG functionality going. This makes interactive discussion and review of interrelated functionality more difficult.
* A couple of times I wanted to "start fresh" to ensure I could replicate a clean install, but with the MEEG functionality. This meant checking out 3 branches after doing a new `git clone`.
* My `GitHub` user account now has repos with very generic names like `data.git` instead of specific ones like `nibabel.git` and `mne-python.git`.
* Having changes in three different places makes it so that some changes need to be merged in e.g. `data` before they can be tested properly on Travis.
* The folder structure is a bit confusing. The code for `reaper` for instance ends up being checked out into `reaper/scitran/reaper`. There must be a way around it because `reaper/code/*` ends up with sub-repos, but it takes more work.
* Right now, some imports to other areas of `scitran` (e.g. from `data` to `api`) must be absolute as `import scitran.data` instead of relative like `from .. import data`. Most packages I see prefer relative imports when possible.

The major advantage I see with the current approach is that it tries to force orthogonality / independence between the individual bits of functionality. Based on my experience, though, the separation actually seems incomplete by necessity/design, so I'm not sure if it buys too much.

**Comment by kofalt** [comment]
Good point on the folder layout; I've implicitly assumed we will refactor the structure while merging as well. In particular `code/api` would become `server`, etc.

We would have to keep the various python interactions in mind, in particular what I mentioned here https://github.com/scitran/scitran/pull/57#issuecomment-131230257 in that I would like to keep the mandatory inputs for the trivial task of pushing JSON around to be as light as possible, keeping data processing (requiring N other libraries) separate. 

Might be something we will have to target after merging repos?

**Comment by gsfr** [comment]
I've always been open to discussion about this, but I've never seen a comprehensive proposal. Mostly just complaints. Sure, there are a lot of legacy reasons here, but also most things are actually working. That's something I'm not willing to sacrifice.

- The reaper currently has a dependency on scitran/data, but really should't. It should be a generic data capture and upload system and thus should remain standalone.
- scitran/data is needed by scitran/api for data parsing, but also by "dcm2nii" for reading and writing. dcm2nii certainly wouldn't need the whole api. Also, the api really shouldn't know what the hell a DICOM is or any other datatype for that matter.

On the flip-side, @rmchamberlain and I have been thinking about how the api might gain the ability to minimally-parse and sort data of all supported types, but without the use of any heavyweight libraries. We can't bloat the api with every possible data library on the planet. With that, full metadata extraction would become a processing job, and those processing containers could depend on specific libs.

Anyway, let's not rush to any conclusions without a comprehensive understanding of all use cases and a solid plan for sustainability as the system expands into further datatypes.

@Eric89GXL, btw, you can easily rename your scitran forks to anything you want, such as scitran-data for scitran/data.

**Comment by Eric89GXL** [comment]
> I've always been open to discussion about this, but I've never seen a comprehensive proposal.

One idea would be to create a single single repository / module `scitran` (i.e., this one) that held all the necessary *submodules* to have a functioning SciTran instance operating on a given machine (developer or deployment). I don't know enough about the repo to know what all modules would need to be subsumed, but how does that sound?

Getting a little bit more specific, but it looks like right now `scitran` more or less pulls in several repos to `scitran/code/` to work. The first thing I'd try would be just putting those bits of code in `scitran/` as submodules, and go from there. I think that would for the most part require lots of path and import modifications. It's possible that I'm underestimating the complexity of the changes that need to be made to reorganize the repo, but maybe that would already be a good start. Even `scitran.reaper` could be in there.

> The reaper currently has a dependency on scitran/data, but really should't.

I think these issue won't be affected too much by whether all the code to run a `scitran` instance is in one repo or multiple. In other repos with separable functionality I've worked on we've implemented simple tests to enforce orthogonality by checking inputs, so it's really not too bad:

https://github.com/vispy/vispy/blob/master/vispy/util/tests/test_import.py

In other words, you ensure that doing `from scitran import reaper` doesn't require `scitran.api` or anything. I wouldn't be surprised if `scipy` had something similar, given that they have pretty good submodule import separation.

> dcm2nii certainly wouldn't need the whole api

That's true, it might only need a subset of the functionality, like `from scitran import api`. As I alluded to above, keep in mind that it is possible in a single-repo setting to keep things orthogonal and lightweight so these operations are not bogged down. For example, `scipy` is a large repo, with some highly independent and some highly interdependent modules. When you do `from scipy import signal`, it doesn't import the `scipy.sparse` or `scipy.ndimage` submodules, but it will import something like `scipy.linalg` if `scipy.signal` needs linear algebra routines (I suspect it might). AFAIK this is how orthogonality plus simpler organization is achieved in most repos. It seems like it should be possible here, too.

Even beyond achieving such independence, my inclination is to have a few extra things in a repo that I don't need to modify to get something working, rather than have to work with multiple (in this case, 3+) repos to add a given bit of functionality. In my mind it complicates the process of adding new things without much benefit.

> most things are actually working. That's something I'm not willing to sacrifice.

I fully agree that any overhaul should preserve existing functionality. In fact, I'd hope that such an overhaul would make things more robust, not less, since it improves the set of tests that can be run. For example, you can look for interactions between all components in the system, which currently would be more difficult to achieve.

**Comment by Eric89GXL** [comment]
Also, if this submodule approach sounds like it might be feasible, I can probably carve out some time in the coming weeks to do some feasibility tests / WIP PR. I'm not sure I'll have all the knowledge to get it working, but it seems like it could be a good opportunity to learn more about how `scitran` works under the hood in any case.

**Comment by gsfr** [comment]
You make a good case, and I'd be interested to see what you can come up with. Any refactoring we do, however, must keep the following concerns in mind and should make progress towards a solution.

Consider that we'll eventually want to handle many different data types. Let's be optimistic and think a few dozen. We started with 3: DICOM (which is huge and far from complete), GE pfiles, and GE physio. Now, we also have MEEG. Already, our external dependencies are considerable. This cannot go on much longer.

- The API should be data type agnostic and thus should not have any data-specific dependencies.
- To enable this, the reaper must become responsible for creating projects, sessions, and acquisitions and for delivering data and metadata to the API in a canonical way. *Sortable* uploads must become a thing of the past, that is, the ```/reaper``` endpoint should go away and with it the API's dependency on ```scitran/data```.
- Each reaper subclass will have certain data-specific dependencies, but running one type of reaper, say DICOM, shouldn't require installation of the dependencies for another type, say MEEG. One solution might be containerizing each reaper.
- Data-specific metadata extraction and processing code, possibly including swaths of external dependencies, must be isolated and ideally containerized.

This list is almost certainly incomplete, but can serve as our starting point.

If the above is achieved, the desirability of having all code in one repo may be diminished, but that's really not my point or concern here.

@lmperry, @rentzso, and I just brainstormed a bit on the whiteboard. Image attached.

![img_0593](https://cloud.githubusercontent.com/assets/712594/9454218/273c6b1c-4a75-11e5-99de-56fa6664c313.JPG)

Cc'ing @rmchamberlain for reference.

**Comment by Eric89GXL** [comment]
Makes sense as far as my limited experience can tell :). I actually went
the other way and put most of the MEEG specific processing in the data PR
-- whoops! That shouldn't be too hard to fix, though.

One thing I'm not sure about is that the data reader has a load_data
method. Is that supposed to work without external dependencies?

**Comment by gsfr** [comment]
```load_data()``` is part of ```scitran/data``` and can thus have arbitrary dependencies. Parts of ```scitran/data``` would be used by the respective processors and possibly by the respective reapers.
